File: light-control/algorithms/_meta/pulse.py
from typing import List
from .pixel import Pixel
from .helpers import bezier_blend
import math

class Pulse:
    def __init__(self, location: float, color: Pixel, size: float, drop_off_right: float, drop_off_left: float) -> None:
        self.location = location
        self.color = color
        self.size = size
        self.drop_off_right = drop_off_right
        self.drop_off_left = drop_off_left
    
    def apply_pulse(self, current_pixels: List[Pixel]) -> List[Pixel]:
        main_bound = self.size / 2.0
        left_bound = max(math.floor(self.location - main_bound - self.drop_off_left), 0);
        right_bound = min(math.ceil(self.location + main_bound + self.drop_off_right), len(current_pixels));
        
        for n in range(left_bound, right_bound + 1):
            pixel = current_pixels[n]
            if n <= self.location:
                if n >= self.location - main_bound:
                    pixel.combine_with(self.color)
                else:
                    distance = self.location - main_bound - n
                    percentage = 1 - (distance / self.drop_off_left)
                    pixel.combine_with(self.color.multiply(bezier_blend(percentage)))
            elif n > self.location:
                if n <= self.location + main_bound:
                    pixel.combine_with(self.color)
                else:
                    distance = n - self.location - main_bound
                    percentage = 1 - (distance / self.drop_off_right)
                    pixel.combine_with(self.color.multiply(bezier_blend(percentage)))


    @property
    def location(self) -> float:
        return self._location
    @location.setter
    def location(self, value: float):
        self._location = value

    @property
    def color(self) -> Pixel:
        return self._color
    @color.setter
    def color(self, value: Pixel):
        self._color = value

    @property
    def size(self) -> float:
        return self._size
    @size.setter
    def size(self, value: float):
        self._size = value
    
    @property
    def drop_off_right(self) -> float:
        return self._drop_off_right
    @drop_off_right.setter
    def drop_off_right(self, value: float):
        self._drop_off_right = value
    
    @property
    def drop_off_left(self) -> float:
        return self._drop_off_left
    @drop_off_left.setter
    def drop_off_left(self, value: float):
        self._drop_off_left = value
    

File: lights-control-server/src/algorithms/_meta/pulse.ts
import { Pixel } from './pixel'
import { bezierBlend } from './helpers'

export class Pulse {
  location: number
  color: Pixel
  size: number
  dropOffRight: number
  dropOffLeft: number

  constructor(
    location: number,
    color: Pixel,
    size: number,
    dropOffRight: number,
    dropOffLeft: number,
  ) {
    this.location = location
    this.color = color
    this.size = size
    this.dropOffRight = dropOffRight
    this.dropOffLeft = dropOffLeft
  }

  applyPulse(pixels: Pixel[]): void {
    const mainBound = this.size / 2.0
    const leftBound = Math.max(Math.floor(this.location - mainBound - this.dropOffLeft), 0)
    const rightBound = Math.min(
      Math.ceil(this.location + mainBound + this.dropOffRight),
      pixels.length - 1,
    )

    for (let n = leftBound; n <= rightBound; n++) {
      const pixel = pixels[n]
      if (n <= this.location) {
        if (n >= this.location - mainBound) {
          pixel.combineWith(this.color)
        } else {
          const distance = this.location - mainBound - n
          const percentage = Math.max(1 - distance / this.dropOffLeft, 0)
          pixel.combineWith(this.color.multiply(bezierBlend(percentage)))
        }
      } else {
        if (n <= this.location + mainBound) {
          pixel.combineWith(this.color)
        } else {
          const distance = n - this.location - mainBound
          const percentage = Math.max(1 - distance / this.dropOffRight, 0)
          pixel.combineWith(this.color.multiply(bezierBlend(percentage)))
        }
      }
    }
  }
}
